sensing/autoware_pointcloud_preprocessor/include/autoware/pointcloud_preprocessor/concatenate_data/cloud_collector.hpp

```diff
@@ -0,0 +1,72 @@
+// Copyright 2024 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "combine_cloud_handler.hpp"
+
+#include <memory>
+#include <string>
+#include <tuple>
+#include <unordered_map>
+
+namespace autoware::pointcloud_preprocessor
+{
+
+class PointCloudConcatenateDataSynchronizerComponent;
+class CombineCloudHandler;
+
+class CloudCollector
+{
+public:
+  CloudCollector(
+    std::shared_ptr<PointCloudConcatenateDataSynchronizerComponent> && ros2_parent_node,
+    std::shared_ptr<CombineCloudHandler> & combine_cloud_handler, int num_of_clouds,
+    double timeout_sec, bool debug_mode);
+
+  // Naive approach
+  void set_arrival_timestamp(double timestamp);
+  double get_arrival_timestamp() const;
+  // Advanced approach
+  void set_reference_timestamp(double timestamp, double noise_window);
+  std::tuple<double, double> get_reference_timestamp_boundary();
+  bool topic_exists(const std::string & topic_name);
+  bool process_pointcloud(
+    const std::string & topic_name, sensor_msgs::msg::PointCloud2::SharedPtr cloud);
+  void concatenate_callback();
+
+  ConcatenatedCloudResult concatenate_pointclouds(
+    std::unordered_map<std::string, sensor_msgs::msg::PointCloud2::SharedPtr> topic_to_cloud_map);
+
+  std::unordered_map<std::string, sensor_msgs::msg::PointCloud2::SharedPtr>
+  get_topic_to_cloud_map();
+
+  bool concatenate_finished() const;
+
+private:
+  std::shared_ptr<PointCloudConcatenateDataSynchronizerComponent> ros2_parent_node_;
+  std::shared_ptr<CombineCloudHandler> combine_cloud_handler_;
+  rclcpp::TimerBase::SharedPtr timer_;
+  std::unordered_map<std::string, sensor_msgs::msg::PointCloud2::SharedPtr> topic_to_cloud_map_;
+  uint64_t num_of_clouds_;
+  double timeout_sec_;
+  double reference_timestamp_min_{0.0};
+  double reference_timestamp_max_{0.0};
+  double arrival_timestamp_{0.0};  // This is used for the naive approach
+  bool debug_mode_;
```

Comment:
This is what I meant with my last comment: the states of two separate and mutually exclusive approaches both end up here. At runtime, only one will be used but both exist, which is
* hard for other developers to understand/maintain
* bugprone if someone does not read this comment

Please refactor (maybe into `CloudCollector` (base class), `CloudCollectorNaive` and `CloudCollectorAdvanced`, or `CloudCollector(const Strategy & strategy)`) and have one common interface for both.
Same goes for `set_arrival_timestamp` and `set_reference_timestamp` and the related member functions above.

Reply:
Thanks for your suggestion, I fixed them in another way in commit https://github.com/autowarefoundation/autoware.universe/pull/8300/commits/1dd7fbc6c70cf2210da66b4619df4c625bfadadf. 

If you think this is not a good approach then let's have some discussion afterward!
Thank you!!!